src/location/util.rs

```diff
@@ -15,51 +15,512 @@
+use mysql::{prelude::Queryable, PooledConn, Row};
 
 use super::{
     ob_part_constants, part_func_type::PartFuncType, ObHashPartDesc, ObKeyPartDesc, ObPartDesc,
     ObPartitionInfo, ObPartitionLevel, ObRangePartDesc, TableEntry,
 };
 use crate::{
+    constant::ALL_DUMMY_TABLE,
     error::{CommonErrCode, Error::Common as CommonErr, Result},
+    location::{
+        ObPartitionEntry, ObPartitionLocation, ObReplicaType, ObServerAddr, ObServerInfo,
+        ObServerRole, ObServerStatus, ReplicaLocation, TableEntryKey, TableLocation, OB_INVALID_ID,
+    },
     rpc::protocol::partition::{
         ob_column::{ObColumn, ObGeneratedColumn, ObSimpleColumn},
         ob_partition_key::{Comparable, ObPartitionKey},
     },
     serde_obkv::value::{CollationLevel, CollationType, ObjMeta, ObjType, Value},
+    util::obversion::{ob_vsn_major, parse_ob_vsn_from_sql},
+    ResultCodes,
 };
 
 pub const TEMPLATE_PART_ID: i32 = -1;
 
-//pub const PROXY_PART_INFO_SQL: &str = "SELECT /*+READ_CONSISTENCY(WEAK)*/
-// part_level, part_num, part_type, part_space, part_expr,    part_range_type,
-// part_interval_bin, interval_start_bin,    sub_part_num, sub_part_type,
-// sub_part_space,    sub_part_range_type, def_sub_part_interval_bin,
-// def_sub_interval_start_bin, sub_part_expr,    part_key_name, part_key_type,
-// part_key_idx, part_key_extra, spare1    FROM oceanbase.
-// __all_virtual_proxy_partition_info    WHERE table_id = {} group by
-// part_key_name order by part_key_name LIMIT {};";
-
 pub struct LocationUtil {}
 
 impl LocationUtil {
-    pub fn fetch_partition_info(
+    pub fn get_ob_version_from_server(conn: &mut my::PooledConn) -> Result<()> {
+        if ob_vsn_major() == 0 {
+            let sql = "SELECT /*+READ_CONSISTENCY(WEAK)*/ OB_VERSION() AS CLUSTER_VERSION";
+            for row in conn.query::<Row, &str>(sql)? {
+                let cluster_version = match my::from_row_opt(row) {
+                    Ok(version) => version,
+                    Err(e) => {
+                        error!("ObTableLocation::get_table_entry_from_remote: fail to do mysql row conversion, err:{}", e);
+                        return Err(CommonErr(
+                            CommonErrCode::ConvertFailed,
+                            format!("mysql row conversion err:{e}"),
+                        ));
+                    }
+                };
+                parse_ob_vsn_from_sql(cluster_version);
+            }
+        }
+        Ok(())
+    }
+
+    /// getTableEntryLocationFromRemote will try get location of every
+    /// partition/tablet from server
+    pub fn get_table_location_from_remote(
         conn: &mut my::PooledConn,
+        key: &TableEntryKey,
+        table_entry: &TableEntry,
+    ) -> Result<ObPartitionEntry> {
+        let partition_num = table_entry.partition_num;
+        let mut part_str = String::with_capacity((partition_num * 7) as usize);
+        let sql: String = if ob_vsn_major() >= 4 {
```

Comment:
Multiple different branches about the ob version occur in this method. Basically, it is not a good style because it's hard to maintain the common logic for the different branches. And compared with so many dynamic branches, one branch at the first with two static methods to handle the case in different ob version looks better.

However, I guess this style or the code content follows the java client's implementation, so the decision is up to you.

Reply:
Fixed, using the function to generate the corresponding SQL